src/DotVVM.Framework.Tests.Common/ResourceManagement/ResourceRepositoryTests.cs

```diff
@@ -25,5 +25,199 @@ public void ResourceRepository_CyclicDependency_Throws()
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "1" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "DotVVM" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "DotVVM" } });
+
+                configuration.Resources.Register("DotVVM", new NullResource { Dependencies = new[] { "7" } }); ;//cyclic
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("7");
+            });
+        }
+
+        [TestMethod]
+        public void ResourceRepository_CyclicDependency_Structure_3_Throws()
+        {
+            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("8", new NullResource { Dependencies = new[] { "7", "6" } });
+
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "5", "4" } });
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "2" } });
+
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "3" } });
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "2" } });
+
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "2", "1" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "5", "1" } });// cyclic
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("8");
+            });
+        }
+
+        [TestMethod]
+        public void ResourceRepository_CyclicDependency_Structure_4_Throws()
+        {
+            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "5" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "4" } });
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "3" } });
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "2" } });
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "1" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "6" } });//cyclic
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("6");
+            });
```

Comment:
How is this test different from `ResourceRepository_CyclicDependency_Structure_2_Throws`?

Reply:
@OndrejSulc:
> 4) Quite a lot I think. ResourceRepository_CyclicDependency_Structure_2_
Throws() has complicated structure but this one is simple tree.

It's not a tree since it has the cycle.